src/gil.rs

```diff
@@ -235,22 +237,26 @@ type PyObjVec = Vec<NonNull<ffi::PyObject>>;
 
 /// Thread-safe storage for objects which were inc_ref / dec_ref while the GIL was not held.
 struct ReferencePool {
+    ever_used: AtomicBool,
     // .0 is INCREFs, .1 is DECREFs
     pointer_ops: sync::Mutex<(PyObjVec, PyObjVec)>,
 }
 
 impl ReferencePool {
     const fn new() -> Self {
         Self {
+            ever_used: AtomicBool::new(false),
             pointer_ops: sync::Mutex::new((Vec::new(), Vec::new())),
         }
     }
 
     fn register_incref(&self, obj: NonNull<ffi::PyObject>) {
+        self.ever_used.store(true, sync::atomic::Ordering::Relaxed);
```

Comment:
It might actually help to even avoid this memory traffic if the flag already is set, e.g.

```rust
self.ever_used.compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed);
```

Reply:
Testing this locally (albeit with a single thread), the `.store` seems to perform better, so will leave as-is.

Reply:
> albeit with a single thread

To be honest, a single-threaded test cannot really tell us anything about the effects of cache line bouncing as there is only a single core's cache involved? (In any case, I would be more keen on the `Lazy`-based approach than open-coding this.)